include/internal/catch_timer.cpp

```diff
@@ -7,67 +7,94 @@
+#include <assert.h>
 
 static const uint64_t nanosecondsInSecond = 1000000000;
+static const uint64_t STD_DEVIATION_THRESHOLD = 100;
 
 namespace Catch {
 
     auto getCurrentNanosecondsSinceEpoch() -> uint64_t {
-        return std::chrono::duration_cast<std::chrono::nanoseconds>( std::chrono::high_resolution_clock::now().time_since_epoch() ).count();
+        return std::chrono::duration_cast<std::chrono::nanoseconds>(
+                std::chrono::high_resolution_clock::now().time_since_epoch()).count();
     }
 
     namespace {
         auto estimateClockResolution() -> uint64_t {
-            uint64_t sum = 0;
-            static const uint64_t iterations = 1000000;
+            uint64_t result = 0;
+            int attempts = 0;
+            do {
+                uint64_t sum = 0;
+                std::vector<uint64_t> deltas;
+                static const uint64_t iterations = 1000000;
+
+                auto startTime = getCurrentNanosecondsSinceEpoch();
 
-            auto startTime = getCurrentNanosecondsSinceEpoch();
+                for (std::size_t i = 0; i < iterations; ++i) {
 
-            for( std::size_t i = 0; i < iterations; ++i ) {
+                    uint64_t ticks;
+                    uint64_t baseTicks = getCurrentNanosecondsSinceEpoch();
+                    do {
+                        ticks = getCurrentNanosecondsSinceEpoch();
+                    } while (ticks == baseTicks);
 
-                uint64_t ticks;
-                uint64_t baseTicks = getCurrentNanosecondsSinceEpoch();
-                do {
-                    ticks = getCurrentNanosecondsSinceEpoch();
-                } while( ticks == baseTicks );
+                    uint64_t delta = ticks - baseTicks;
+                    deltas.push_back(delta);
+                    sum += delta;
+                    // If we have been calibrating for over 3 seconds -- the clock
+                    // is terrible and we should move on.
+                    // TBD: How to signal that the measured resolution is probably wrong?
+                    if (ticks > startTime + 3 * nanosecondsInSecond) {
+                        return sum / (i + 1u);
+                    }
 
-                auto delta = ticks - baseTicks;
-                sum += delta;
+                    // We're just taking the mean, here. To do better we could take the std. dev and exclude outliers
+                    // - and potentially do more iterations if there's a high variance.
+                    result = sum / iterations;
+                    uint64_t squared_difference_sum = 0;
+                    for (std::size_t iterator = 0; iterator < iterations; ++iterator) {
+                        squared_difference_sum += (deltas[iterator] - result) * (deltas[iterator] - result);
+                    }
+                    long double std_deviation = static_cast<long double>(squared_difference_sum) / iterations;
+                    if (std_deviation < STD_DEVIATION_THRESHOLD) {
```

Comment:
This doesn't look right -- you do this check after every iteration, which means that it will return after the first measurement iteration, and use whatever was measured in the first iteration as the resolution estimate.

Shouldn't it be on the outside of this loop?

Reply:
Open up a new one then.